src/v/cluster/rm_stm.cc

```diff
@@ -0,0 +1,1008 @@
+
+        vlog(
+          clusterlog.error,
+          "Set fence {} is lesser that prepare's epoch: {}",
+          fence_it->second,
+          pid);
+        if (_recovery_policy == best_effort) {
+            fence_it->second = epoch(pid);
+        } else {
+            vassert(
+              false,
+              "Set fence {} is lesser that prepare's epoch: {}",
+              fence_it->second,
+              pid);
+        }
+    } else if (fence_it->second > epoch(pid)) {
+        // during a data race between replicating
+        //  - a prepare record and
+        //  - a record wth higher epoch for the same pid.id e.g abort
+        // the latter request won and prepare got fenced off
+
+        // cleaning the state since this pid is fenced off
+        vlog(
+          clusterlog.warn,
+          "Prepare with pid:{} is fenced off by {}",
+          pid,
+          fence_it->second);
+        _mem_state.expected.erase(pid);
+        return;
+    }
+
+    if (_log_state.aborted.contains(pid)) {
+        // during a data race between replicating a prepare record
+        // and an abort record, the abort record won
+        vlog(clusterlog.warn, "Can't prepare an aborted tx (pid:{})", pid);
+        _mem_state.expected.erase(pid);
+        return;
+    }
+
+    _log_state.prepared.try_emplace(pid, prepare);
+    _mem_state.expected.erase(pid);
+
+    auto has_applied_it = _mem_state.has_prepare_applied.find(pid);
+    if (has_applied_it != _mem_state.has_prepare_applied.end()) {
+        // _mem_state.has_prepare_applied may lack pid if
+        // this is processing og the historic prepare
+        has_applied_it->second = true;
+    }
+}
+
+void rm_stm::apply_control(
+  model::producer_identity pid, model::control_record_type crt) {
+    auto fence_it = _log_state.fence_pid_epoch.find(id(pid));
+    if (fence_it == _log_state.fence_pid_epoch.end()) {
+        _log_state.fence_pid_epoch.emplace(id(pid), epoch(pid));
+    } else if (fence_it->second < epoch(pid)) {
+        fence_it->second = epoch(pid);
+    }
+
+    // either epoch is the same as fencing or it's lesser in the latter
+    // case we don't fence off aborts and commits because transactional
+    // manager already decided a tx's outcome and acked it to the client
+
+    if (crt == model::control_record_type::tx_abort) {
+        if (_log_state.aborted.contains(pid)) {
+            // already aborted; but it's fine - a transaction
+            // coordinator may re(abort) a transaction on recovery
+            // or during concurrent init_tx
+            return;
+        }
+
+        _log_state.prepared.erase(pid);
+        auto offset_it = _log_state.ongoing_map.find(pid);
+        if (offset_it != _log_state.ongoing_map.end()) {
+            _log_state.aborted.emplace(pid, offset_it->second);
+            _log_state.ongoing_set.erase(offset_it->second.first);
+            _log_state.ongoing_map.erase(pid);
+        }
```

Comment:
There's a lot going on here... is it worth pulling this out into a data structure that encapsulates it all?